frame/bounties/src/lib.rs

```diff
@@ -660,12 +785,685 @@ decl_module! {
+						curator_deposit: &mut BalanceOf<T>| {
+							let imbalance = T::Currency::slash_reserved(
+								arg_curator,
+								*curator_deposit,
+							).0;
+							T::OnSlash::on_unbalanced(imbalance);
+							*curator_deposit = Zero::zero();
+						};
+
+					match subbounty.status {
+						SubBountyStatus::Added => {
+							// No curator to unassign at this point.
+							return Err(Error::<T>::UnexpectedStatus.into())
+						}
+						SubBountyStatus::SubCuratorProposed { ref subcurator } => {
+							// A subcurator has been proposed, but not accepted yet.
+							// Either `RejectOrigin`, curator or the proposed subcurator
+							// can unassign the subcurator.
+							ensure!(
+								maybe_sender.map_or(
+									true,
+									|sender| sender == *subcurator ||
+										maybe_bounty_active_status.map_or(false, |v| v.0 == sender)
+								),
+								BadOrigin,
+							);
+
+						},
+						SubBountyStatus::Active { ref subcurator } => {
+							// The bounty is active.
+							match maybe_sender {
+								// If the `RejectOrigin` is calling this function,
+								// slash the subcurator deposit.
+								None => {
+									slash_curator(subcurator, &mut subbounty.curator_deposit);
+									// Continue to change bounty status below...
+								},
+								Some(sender) => {
+									if sender == *subcurator {
+										// This is the subcurator,
+										// willingly giving up their role.
+										// Give back their deposit.
+										T::Currency::unreserve(
+											&subcurator,
+											subbounty.curator_deposit,
+										);
+										// Continue to change bounty status below...
+									} else if maybe_bounty_active_status
+										.as_ref()
+										.expect("Status is none iff origin is none; qed").0 == sender
+									{
+										// looks like subcurator is inactive,
+										// slash the subcurator deposit.
+										slash_curator(subcurator, &mut subbounty.curator_deposit);
+										// Continue to change bounty status below...
+									} else {
+										// check for expiry
+										// looks like subcurator is inactive,
+										// slash the subcurator deposit.
+										let block_number = system::Pallet::<T>::block_number();
+										if maybe_bounty_active_status
+											.as_ref()
+											.expect("Status is none if origin is none; qed").1 < block_number
+										{
+											slash_curator(
+												subcurator,
+												&mut subbounty.curator_deposit,
+											);
+											// Continue to change bounty status below...
+										} else {
+											// Curator has more time to give an update.
+											return Err(Error::<T>::Premature.into())
+										}
+									}
+								},
+							}
+						},
+						SubBountyStatus::PendingPayout { ref subcurator, .. } => {
+							ensure!(
+								maybe_sender.map_or(
+									true,
+									|sender| maybe_bounty_active_status.map_or(true, |v| v.0 == sender),
+								),
+								BadOrigin,
+							);
```

Comment:
This chunk is repeated and could be turned into a helper.

Reply:
Status?

Reply:
Sure Kian Will pick up the refactoring
.